Lesson20/feyza/oop.js

```diff
@@ -0,0 +1,174 @@
+//Override the describe() method in the Square class to log a description of the square to the console (e.g. 
+//"This is a blue square with a side length of 5 and an area of 25").
+
+class Shape {
+    constructor(name, color) {
+        this.name = name;
+        this.color = color;
+    }
+
+    describe() {
+        console.log(`This is a ${this.color} ${this.name}.`);
+    }
+}
+
+class Square extends Shape {
+    constructor(name, color, sideLength) {
+        super(name, color);
+        this.sideLength = sideLength;
+        this.area = sideLength * sideLength;
+    }
+
+    describe() {
+        console.log(`This is a ${this.color} ${this.name} with a side length of ${this.sideLength} and an area of ${this.area}.`)
+    }
+}
+
+let shape1 = new Shape("triangle", "purple");
+let square1 = new Square("kare", "yellow", 5);
+
+console.log(shape1);
+console.log(square1);
+
+shape1.describe();
+square1.describe();
+
+
+// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that 
+//the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in 
+//the Car class to log a different message to the console indicating that the car has started.
+
+class Vehicle {
+    constructor(make, model, year) {
+        this.make = make;
+        this.model = model;
+        this.year = year;
+    }
+
+    start() {
+        console.log(`The vehicle has started : ${this.make} ${this.model} ${this.year}`);
+    }
+}
+
+class Car extends Vehicle {
+    constructor(make, model, year, numDoor) {
+        super(make, model, year);
+        this.numDoor = numDoor;
+    }
+
+    start() {
+        console.log(`The car has started : ${this.make} ${this.model} ${this.year} with ${this.numDoor} doors.`);
+    }
+}
+
+let vehicle1 = new Vehicle("Audi", "Q7", 2022);
+let car1 = new Car("Kia", "EV9", 2024, 4);
+
+console.log(vehicle1);
+console.log(car1);
+
+vehicle1.start();
+car1.start();
+
+// 4. Create a BankAccount class with balance and interestRate properties, and deposit() and withdraw() methods. Add a transactionHistory property that is 
+//an array of objects representing each transaction made on the account, with properties for type (either "deposit" or "withdrawal"), amount, 
+//and date ({type: "withdrawal", amount: amount, date: new Date().toLocaleDateString()}). Implement deposit() and withdraw() methods that add new transactions to
+// the transactionHistory array. Implement a get transactionHistory() getter method that returns a copy of the transaction history array, and a get 
+//currentBalance() getter method that calculates and returns the current balance of the account based on the transaction history.
+
+class BankAccount {
+    Constructor(balance, interestRate) {
+        this.balance = balance;
+        this.interestRate = interestRate;
```

Comment:
The class should also have transactionHistory property 
```suggestion
        this.transactionHistory = [];
```